File: src/lib/recommendation-log.ts
"use client";

import type { TrackInfo } from "./tauri";
import { cache } from "./tauri";

const KEY = "recommendation_log_v1";
const MAX_EVENTS = 800;
const DAY_S = 24 * 3600;

export type RecommendationSource = "pet" | "auto" | "radio" | "search";

export type RecommendationEvent = {
  trackId: number;
  ts: number;
  source: RecommendationSource;
  /**
   * 推荐时的主艺人名 —— v2 新增,用来做 artist-level fatigue。
   * 老事件没这个字段,读取时按缺失处理。
   */
  artist?: string;
};

export type RecentRecommendationContext = {
  last24hTrackIds: Set<number>;
  last7dTrackIds: Set<number>;
  /** 近 24h 每个艺人(normalize 后)被推荐的次数 */
  last24hArtistCounts: Map<string, number>;
  /** 近 7d 同上 */
  last7dArtistCounts: Map<string, number>;
};

/**
 * artist key:跟 track-dedupe/pet-agent 的 normalizeArtistKey 等价 —— 全部小写,
 * 去掉常见标点。不引入 track-dedupe 的依赖(那边还涉及 song title 规范),
 * 这里只关心 artist,保持本模块自包含。
 */
export function normalizeArtistKeyForLog(name: string): string {
  return name
    .toLowerCase()
    .replace(/[\s'"`·・\-－—_,，。.、!?！？]+/g, "");
}

let buffer: RecommendationEvent[] | null = null;
let writeTimer: number | null = null;

async function ensureBuffer(): Promise<RecommendationEvent[]> {
  if (buffer) return buffer;
  try {
    const raw = await cache.getState(KEY);
    if (raw) {
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed)) {
        buffer = parsed as RecommendationEvent[];
        return buffer;
      }
    }
  } catch {
    /* 坏缓存忽略 */
  }
  buffer = [];
  return buffer;
}

function scheduleFlush() {
  if (writeTimer != null) window.clearTimeout(writeTimer);
  writeTimer = window.setTimeout(() => {
    writeTimer = null;
    if (!buffer) return;
    const trimmed = buffer.slice(-MAX_EVENTS);
    if (trimmed.length !== buffer.length) buffer = trimmed;
    cache.setState(KEY, JSON.stringify(buffer)).catch((e) => {
      console.debug("[pipo] recommendation_log 落盘失败", e);
    });
  }, 50);
}

/**
 * 落事件 —— 支持两种入参形式:
 *   - number[] 旧调用:不带艺人,fatigue 算不到只能按 trackId 推断不出来
 *   - TrackInfo[] 推荐用形式:把主艺人写进事件,后续可以按 artist 聚合
 *
 * 两种形式并存是为了不破坏现有 caller。新代码请尽量传 TrackInfo[]。
 */
export async function logRecommendations(
  tracks: number[] | TrackInfo[],
  source: RecommendationSource,
): Promise<void> {
  if (tracks.length === 0) return;
  const buf = await ensureBuffer();
  const now = Math.floor(Date.now() / 1000);
  const seen = new Set<number>();
  for (const item of tracks) {
    if (typeof item === "number") {
      if (seen.has(item)) continue;
      seen.add(item);
      buf.push({ trackId: item, ts: now, source });
    } else {
      if (seen.has(item.id)) continue;
      seen.add(item.id);
      const artist = item.artists[0]?.name;
      buf.push({ trackId: item.id, ts: now, source, artist });
    }
  }
  scheduleFlush();
}

export async function readRecommendationLog(): Promise<RecommendationEvent[]> {
  return ensureBuffer().then((b) => b.slice());
}

export async function readRecentRecommendationContext(): Promise<RecentRecommendationContext> {
  const events = await readRecommendationLog();
  const now = Math.floor(Date.now() / 1000);
  const last24hTrackIds = new Set<number>();
  const last7dTrackIds = new Set<number>();
  const last24hArtistCounts = new Map<string, number>();
  const last7dArtistCounts = new Map<string, number>();

  for (const ev of events) {
    const ageS = now - ev.ts;
    const in7d = ageS <= 7 * DAY_S;
    const in24h = ageS <= DAY_S;
    if (in24h) last24hTrackIds.add(ev.trackId);
    if (in7d) last7dTrackIds.add(ev.trackId);

    if (!ev.artist) continue; // 老事件没 artist 字段,跳过 artist 聚合
    const key = normalizeArtistKeyForLog(ev.artist);
    if (!key) continue;
    if (in24h) {
      last24hArtistCounts.set(key, (last24hArtistCounts.get(key) ?? 0) + 1);
    }
    if (in7d) {
      last7dArtistCounts.set(key, (last7dArtistCounts.get(key) ?? 0) + 1);
    }
  }

  return {
    last24hTrackIds,
    last7dTrackIds,
    last24hArtistCounts,
    last7dArtistCounts,
  };
}

File: android-native/app/src/main/java/app/pipo/nativeapp/data/RecommendationLog.kt
package app.pipo.nativeapp.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * AI 推荐日志 —— 镜像 src/lib/recommendation-log.ts。
 *
 * 让 AI 知道"24h 内推过的歌别再推"，避免 candidate-recall 重复同一批结果。
 * 每次 AI 出了 batch 都 log 一次（source=pet）；自动播放推荐用 source=auto；电台 radio。
 */
class RecommendationLog(context: Context) {
    enum class Source(val key: String) { Pet("pet"), Auto("auto"), Radio("radio"), Search("search") }

    data class Event(
        val trackId: Long,
        val tsSec: Long,
        val source: Source,
        /** 主艺人名 —— v2 新增,做 artist-level fatigue;老事件没这字段当 null */
        val artist: String? = null,
    )

    data class RecentContext(
        val last24hTrackIds: Set<Long>,
        val last7dTrackIds: Set<Long>,
        /** 近 24h 每个艺人(normalize 后)被推荐的次数 */
        val last24hArtistCounts: Map<String, Int> = emptyMap(),
        /** 近 7d 同上 */
        val last7dArtistCounts: Map<String, Int> = emptyMap(),
    )

    private val prefs = context.applicationContext
        .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    @Volatile
    private var buffer: MutableList<Event>? = null

    @Synchronized
    private fun ensureBuffer(): MutableList<Event> {
        buffer?.let { return it }
        val raw = prefs.getString(KEY, null)
        val out = mutableListOf<Event>()
        if (raw != null) {
            runCatching {
                val arr = JSONArray(raw)
                for (i in 0 until arr.length()) {
                    val o = arr.optJSONObject(i) ?: continue
                    val srcKey = o.optString("source")
                    val src = Source.entries.firstOrNull { it.key == srcKey } ?: Source.Pet
                    val artistField = o.optString("artist", "")
                    val artist = if (artistField.isBlank()) null else artistField
                    out.add(Event(o.optLong("trackId"), o.optLong("ts"), src, artist))
                }
            }
        }
        buffer = out
        return out
    }

    @Synchronized
    fun log(trackIds: List<Long>, source: Source = Source.Pet) {
        if (trackIds.isEmpty()) return
        val buf = ensureBuffer()
        val now = System.currentTimeMillis() / 1000
        val seen = HashSet<Long>()
        for (id in trackIds) {
            if (!seen.add(id)) continue
            buf.add(Event(id, now, source))
        }
        flush()
    }

    /**
     * v2:支持传 NativeTrack[],把主艺人名一起写进日志,后续 fatigue 才能按艺人聚合。
     * 新代码请尽量用这个;老的 log(List<Long>) 保留兼容。
     */
    @Synchronized
    fun logTracks(tracks: List<NativeTrack>, source: Source = Source.Pet) {
        if (tracks.isEmpty()) return
        val buf = ensureBuffer()
        val now = System.currentTimeMillis() / 1000
        val seen = HashSet<Long>()
        for (t in tracks) {
            val id = t.neteaseId ?: continue
            if (!seen.add(id)) continue
            val artist = t.artist.split('/', '&', ',', '、').firstOrNull()?.trim()
            buf.add(Event(id, now, source, artist))
        }
        flush()
    }

    @Synchronized
    private fun flush() {
        val buf = buffer ?: return
        val trimmed = if (buf.size > MAX_EVENTS) buf.takeLast(MAX_EVENTS).toMutableList() else buf
        buffer = trimmed
        val arr = JSONArray()
        trimmed.forEach {
            arr.put(JSONObject().apply {
                put("trackId", it.trackId); put("ts", it.tsSec); put("source", it.source.key)
                if (!it.artist.isNullOrBlank()) put("artist", it.artist)
            })
        }
        prefs.edit().putString(KEY, arr.toString()).apply()
    }

    fun readAll(): List<Event> = ensureBuffer().toList()

    fun recentContext(): RecentContext {
        val events = readAll()
        val now = System.currentTimeMillis() / 1000
        val day = mutableSetOf<Long>()
        val week = mutableSetOf<Long>()
        val dayArtist = HashMap<String, Int>()
        val weekArtist = HashMap<String, Int>()
        for (e in events) {
            val age = now - e.tsSec
            val in7d = age <= 7 * DAY_S
            val in24h = age <= DAY_S
            if (in24h) day.add(e.trackId)
            if (in7d) week.add(e.trackId)
            val artist = e.artist ?: continue
            val key = normalizeArtistKey(artist)
            if (key.isEmpty()) continue
            if (in24h) dayArtist[key] = (dayArtist[key] ?: 0) + 1
            if (in7d) weekArtist[key] = (weekArtist[key] ?: 0) + 1
        }
        return RecentContext(day, week, dayArtist, weekArtist)
    }

    private fun normalizeArtistKey(s: String): String =
        s.lowercase().replace(Regex("[\\s'\"`·・\\-－—_,，。.、!?！？]+"), "")

    companion object {
        private const val PREFS_NAME = "claudio_recommendation_log"
        private const val KEY = "v1"
        private const val MAX_EVENTS = 800
        private const val DAY_S = 24 * 3600L
    }
}
